### ISSUE TITLE
A puzzle about gen_update function

### ISSUE THREAD

**Issue by ssdutHB** [mingyuliutw/UNIT#22]
<pre style="font-family: 宋体; font-size: 14.4pt;"><span style="color:#000080;font-weight:bold;">def </span>gen_update(<span style="color:#94558d;">self</span>, images_a, images_b, hyperparameters):<br>  <span style="color:#94558d;">self</span>.gen.zero_grad()<br>  x_aa, x_ba, x_ab, x_bb, shared = <span style="color:#94558d;">self</span>.gen(images_a, images_b)<br>  x_bab, shared_bab = <span style="color:#94558d;">self</span>.gen.forward_a2b(x_ba)<br>  x_aba, shared_aba = <span style="color:#94558d;">self</span>.gen.forward_b2a(x_ab)<br>  outs_a, outs_b = <span style="color:#94558d;">self</span>.dis(x_ba,x_ab)<br>  <span style="color:#000080;font-weight:bold;">for </span>it, (out_a, out_b) <span style="color:#000080;font-weight:bold;">in </span><span style="color:#000080;">enumerate</span>(itertools.izip(outs_a, outs_b)):<br>    outputs_a = nn.functional.sigmoid(out_a)<br>    outputs_b = nn.functional.sigmoid(out_b)<br>    all_ones = Variable(torch.ones((outputs_a.size(<span style="color:#0000ff;">0</span>))).cuda(<span style="color:#94558d;">self</span>.gpu))<br>    <span style="color:#000080;font-weight:bold;">if </span>it==<span style="color:#0000ff;">0</span>:<br>      ad_loss_a = nn.functional.binary_cross_entropy(outputs_a, all_ones)<br>      ad_loss_b = nn.functional.binary_cross_entropy(outputs_b, all_ones)<br>    <span style="color:#000080;font-weight:bold;">else</span>:<br>      ad_loss_a += nn.functional.binary_cross_entropy(outputs_a, all_ones)<br>      ad_loss_b += nn.functional.binary_cross_entropy(outputs_b, all_ones)</pre>

The code above is a part of code in cocogan_trainer.py.
 I think the 
` all_ones = Variable(torch.ones((outputs_a.size(0))).cuda(self.gpu))`
should be 
` all_zeros = Variable(torch.zeros((outputs_a.size(0))).cuda(self.gpu))`
Because it calculates the loss when the inputs of Discriminator are fakeA and fakeB.
Is my understanding right? 
Do I misunderstand it?

**Comment by mingyuliutw** [comment]
Intuitively, the goal of the generator is to deceive the discriminator. Hence, it needs to be updated in a way the discriminator will output ones for the generated images. Instead of optimizing the original GAN objective, the way it is implemented in the release leads to more stable training. This is a common way of implementing the GAN learning algorithm. You could check out Goodfellow et al.'s NIPS'14 paper to see the description.

**Comment by ssdutHB** [comment]
Thank you very much. I get it!
